src/groupeddataframe/splitapplycombine.jl

```diff
@@ -523,58 +541,66 @@ function _combine(gd::GroupedDataFrame,
     # and if a given column can be replaced in the future
     seen_cols = Dict{Symbol, Tuple{Bool, Int}}()
 
+    tasks = similar(cs_norm, Task)
+
     parentdf = parent(gd)
-    for i in eachindex(cs_norm, optional_transform)
+    for i in eachindex(cs_norm, optional_transform, tasks)
         cs_i = cs_norm[i]
         optional_i = optional_transform[i]
 
-        if length(gd) > 0 && isagg(cs_i, gd)
+        tasks[i] = Threads.@spawn if length(gd) > 0 && isagg(cs_i, gd)
             _combine_process_agg(cs_i, optional_i, parentdf, gd, seen_cols, trans_res, idx_agg)
         elseif keeprows && cs_i isa Pair && first(last(cs_i)) === identity &&
                !(first(cs_i) isa AsTable) && (last(last(cs_i)) isa Symbol)
             # this is a fast path used when we pass a column or rename a column in select or transform
             _combine_process_noop(cs_i, optional_i, parentdf, seen_cols, trans_res, idx_keeprows, copycols)
         elseif cs_i isa Base.Callable
-            idx_callable = _combine_process_callable(cs_i, optional_i, parentdf, gd, seen_cols, trans_res, idx_agg)
-            if idx_callable !== nothing
-                if idx_agg === nothing
-                    idx_agg = idx_callable
-                else
-                    @assert idx_agg === idx_callable
-                end
-            end
+            _combine_process_callable(cs_i, optional_i, parentdf, gd, seen_cols, trans_res, idx_agg)
         else
             @assert cs_i isa Pair
-            idx_pair = _combine_process_pair(cs_i, optional_i, parentdf, gd, seen_cols, trans_res, idx_agg)
-            if idx_pair !== nothing
-                if idx_agg === nothing
-                    idx_agg = idx_pair
-                else
-                    @assert idx_agg === idx_pair
-                end
+            _combine_process_pair(cs_i, optional_i, parentdf, gd, seen_cols, trans_res, idx_agg)
+        end
+    end
+    for t in tasks
+        local postprocessf
+        try
```

Comment:
what is the purpose of this `try`? Just to unwrap one level of stack trace?

Reply:
That's to avoid throwing a `TaskFailedException`. Without it:
```julia
julia> combine(gd, :y => (x -> nonexistent(x)))
ERROR: TaskFailedException
Stacktrace:
 [1] wait
   @ ./task.jl:317 [inlined]
 [2] fetch
   @ ./task.jl:332 [inlined]
 [3] _combine(gd::GroupedDataFrame{DataFrame}, cs_norm::Vector{Any}, optional_transform::Vector{Bool}, copycols::Bool, keeprows::Bool, renamecols::Bool)
   @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:604
 [4] _combine_prepare(gd::GroupedDataFrame{DataFrame}, cs::Union{Regex, AbstractString, Function, Signed, Symbol, Unsigned, Pair, AbstractVector{T} where T, Type, All, Between, Cols, InvertedIndex}; keepkeys::Bool, ungroup::Bool, copycols::Bool, keeprows::Bool, renamecols::Bool)
   @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:66
 [5] #combine#519
   @ ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:703 [inlined]
 [6] combine(gd::GroupedDataFrame{DataFrame}, cs::Pair{Symbol, var"#17#18"})
   @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:696
 [7] top-level scope
   @ REPL[14]:1

    nested task error: UndefVarError: nonexistent not defined
    Stacktrace:
     [1] (::var"#17#18")(x::SubArray{Float64, 1, Vector{Float64}, Tuple{Vector{Int64}}, false})
       @ Main ./REPL[14]:1
     [2] do_call(f::var"#17#18", idx::Vector{Int64}, starts::Vector{Int64}, ends::Vector{Int64}, gd::GroupedDataFrame{DataFrame}, incols::Tuple{Vector{Float64}}, i::Int64)
       @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/callprocessing.jl:95
     [3] _combine_process_pair(cs_i::Pair, optional_i::Bool, parentdf::DataFrame, gd::GroupedDataFrame{DataFrame}, seen_cols::Dict{Symbol, Tuple{Bool, Int64}}, trans_res::Vector{DataFrames.TransformationResult}, idx_agg::Base.RefValue{Union{Nothing, Vector{Int64}}})
       @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:508
     [4] macro expansion
       @ ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:598 [inlined]
     [5] (::DataFrames.var"#806#812"{GroupedDataFrame{DataFrame}, Bool, Bool, DataFrame, Dict{Symbol, Tuple{Bool, Int64}}, Vector{DataFrames.TransformationResult}, Base.RefValue{Union{Nothing, Vector{Int64}}}, Bool, Pair{Int64, Pair{var"#17#18", Symbol}}})()
       @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:541
```

With it:
```julia
julia> combine(gd, :y => (x -> nonexistent(x)))
ERROR: UndefVarError: nonexistent not defined
Stacktrace:
 [1] wait
   @ ./task.jl:317 [inlined]
 [2] fetch
   @ ./task.jl:332 [inlined]
 [3] _combine(gd::GroupedDataFrame{DataFrame}, cs_norm::Vector{Any}, optional_transform::Vector{Bool}, copycols::Bool, keeprows::Bool, renamecols::Bool)
   @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:604
 [4] _combine_prepare(gd::GroupedDataFrame{DataFrame}, cs::Union{Regex, AbstractString, Function, Signed, Symbol, Unsigned, Pair, AbstractVector{T} where T, Type, All, Between, Cols, InvertedIndex}; keepkeys::Bool, ungroup::Bool, copycols::Bool, keeprows::Bool, renamecols::Bool)
   @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:66
 [5] #combine#519
   @ ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:703 [inlined]
 [6] combine(gd::GroupedDataFrame{DataFrame}, cs::Pair{Symbol, var"#19#20"})
   @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:703
 [7] top-level scope
   @ REPL[14]:1
```

Another option would be to use `throw(t.exception)` instead of `rethrow(t.exception)`. The gives three exceptions, which is too bad, but the last one gives the line where the error really happened. But we probably don't care since that's not useful for users. Ideally we would only print the last one but I don't know how to do that.
```julia
julia> combine(gd, :y => (x -> nonexistent(x)))
ERROR: UndefVarError: nonexistent not defined
Stacktrace:
 [1] _combine(gd::GroupedDataFrame{DataFrame}, cs_norm::Vector{Any}, optional_transform::Vector{Bool}, copycols::Bool, keeprows::Bool, renamecols::Bool)
   @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:607
 [2] _combine_prepare(gd::GroupedDataFrame{DataFrame}, cs::Union{Regex, AbstractString, Function, Signed, Symbol, Unsigned, Pair, AbstractVector{T} where T, Type, All, Between, Cols, InvertedIndex}; keepkeys::Bool, ungroup::Bool, copycols::Bool, keeprows::Bool, renamecols::Bool)
   @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:66
 [3] #combine#519
   @ ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:703 [inlined]
 [4] combine(gd::GroupedDataFrame{DataFrame}, cs::Pair{Symbol, var"#21#22"})
   @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:703
 [5] top-level scope
   @ REPL[14]:1

caused by: TaskFailedException
Stacktrace:
 [1] wait
   @ ./task.jl:317 [inlined]
 [2] fetch
   @ ./task.jl:332 [inlined]
 [3] _combine(gd::GroupedDataFrame{DataFrame}, cs_norm::Vector{Any}, optional_transform::Vector{Bool}, copycols::Bool, keeprows::Bool, renamecols::Bool)
   @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:604
 [4] _combine_prepare(gd::GroupedDataFrame{DataFrame}, cs::Union{Regex, AbstractString, Function, Signed, Symbol, Unsigned, Pair, AbstractVector{T} where T, Type, All, Between, Cols, InvertedIndex}; keepkeys::Bool, ungroup::Bool, copycols::Bool, keeprows::Bool, renamecols::Bool)
   @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:66
 [5] #combine#519
   @ ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:703 [inlined]
 [6] combine(gd::GroupedDataFrame{DataFrame}, cs::Pair{Symbol, var"#21#22"})
   @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:703
 [7] top-level scope
   @ REPL[14]:1

    nested task error: UndefVarError: nonexistent not defined
    Stacktrace:
     [1] (::var"#21#22")(x::SubArray{Float64, 1, Vector{Float64}, Tuple{Vector{Int64}}, false})
       @ Main ./REPL[14]:1
     [2] do_call(f::var"#21#22", idx::Vector{Int64}, starts::Vector{Int64}, ends::Vector{Int64}, gd::GroupedDataFrame{DataFrame}, incols::Tuple{Vector{Float64}}, i::Int64)
       @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/callprocessing.jl:95
     [3] _combine_process_pair(cs_i::Pair, optional_i::Bool, parentdf::DataFrame, gd::GroupedDataFrame{DataFrame}, seen_cols::Dict{Symbol, Tuple{Bool, Int64}}, trans_res::Vector{DataFrames.TransformationResult}, idx_agg::Base.RefValue{Union{Nothing, Vector{Int64}}})
       @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:508
     [4] macro expansion
       @ ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:598 [inlined]
     [5] (::DataFrames.var"#940#946"{GroupedDataFrame{DataFrame}, Bool, Bool, DataFrame, Dict{Symbol, Tuple{Bool, Int64}}, Vector{DataFrames.TransformationResult}, Base.RefValue{Union{Nothing, Vector{Int64}}}, Bool, Pair{Int64, Pair{var"#21#22", Symbol}}})()
       @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:541
```

Reply:
The issue with the current `rethrow` approach is that it hides all the actual stacktrace, for example if you do:
```julia
julia> f(x) = nonexistent(x)
f (generic function with 1 method)

julia> combine(gd, :y => (x -> f(x)))
ERROR: UndefVarError: nonexistent not defined
Stacktrace:
 [1] wait
   @ ./task.jl:317 [inlined]
 [2] fetch
   @ ./task.jl:332 [inlined]
 [3] _combine(gd::GroupedDataFrame{DataFrame}, cs_norm::Vector{Any}, optional_transform::Vector{Bool}, copycols::Bool, keeprows::Bool, renamecols::Bool)
   @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:604
 [4] _combine_prepare(gd::GroupedDataFrame{DataFrame}, cs::Union{Regex, AbstractString, Function, Signed, Symbol, Unsigned, Pair, AbstractVector{T} where T, Type, All, Between, Cols, InvertedIndex}; keepkeys::Bool, ungroup::Bool, copycols::Bool, keeprows::Bool, renamecols::Bool)
   @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:66
 [5] #combine#519
   @ ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:703 [inlined]
 [6] combine(gd::GroupedDataFrame{DataFrame}, cs::Pair{Symbol, var"#29#30"})
   @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:703
 [7] top-level scope
   @ REPL[18]:1
```
That can be really problematic as it makes debugging impossible. Not sure whether there's a way to print only the error we want to show to users.

Reply:
Since with your approach:
```
_combine(gd::GroupedDataFrame{DataFrame}, cs_norm::Vector{Any}, optional_transform::Vector{Bool}, copycols::Bool, keeprows::Bool, renamecols::Bool)
```
is still printed (and it is a crucial line) I am OK to have it as you proposed. But I would be also OK not to do the `try-catch` block and just leave what is printed by default.

Reply:
I wouldn't say that's the crucial line: the user doesn't care where the error happened in DataFrames, since that's not an error in DataFrames. Users care about where in their code the error happened.

I've pushed a commit to just use the default error handling. This gives something a bit verbose but with all the information. What's annoying is that the first exception is `TaskFailedException`, so I've added a comment at https://github.com/JuliaLang/julia/pull/37224#issuecomment-743804791.
```julia
julia> f(x) = nonexistent(x)
f (generic function with 1 method)

julia> combine(gd, :x => (x -> f(x)))
ERROR: TaskFailedException
Stacktrace:
 [1] wait
   @ ./task.jl:317 [inlined]
 [2] fetch
   @ ./task.jl:332 [inlined]
 [3] _combine(gd::GroupedDataFrame{DataFrame}, cs_norm::Vector{Any}, optional_transform::Vector{Bool}, copycols::Bool, keeprows::Bool, renamecols::Bool)
   @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:562
 [4] _combine_prepare(gd::GroupedDataFrame{DataFrame}, cs::Union{Regex, AbstractString, Function, Signed, Symbol, Unsigned, Pair, AbstractVector{T} where T, Type, All, Between, Cols, InvertedIndex}; keepkeys::Bool, ungroup::Bool, copycols::Bool, keeprows::Bool, renamecols::Bool)
   @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:66
 [5] #combine#529
   @ ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:645 [inlined]
 [6] combine(gd::GroupedDataFrame{DataFrame}, cs::Pair{Symbol, var"#21#22"})
   @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:647
 [7] top-level scope
   @ REPL[34]:1

    nested task error: UndefVarError: nonexistent not defined
    Stacktrace:
     [1] f(x::SubArray{Int64, 1, Vector{Int64}, Tuple{Vector{Int64}}, false})
       @ Main ./REPL[6]:1
     [2] #21
       @ ./REPL[34]:1 [inlined]
     [3] do_call(f::var"#21#22", idx::Vector{Int64}, starts::Vector{Int64}, ends::Vector{Int64}, gd::GroupedDataFrame{DataFrame}, incols::Tuple{Vector{Int64}}, i::Int64)
       @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/callprocessing.jl:95
     [4] _combine_process_pair(cs_i::Pair, optional_i::Bool, parentdf::DataFrame, gd::GroupedDataFrame{DataFrame}, seen_cols::Dict{Symbol, Tuple{Bool, Int64}}, trans_res::Vector{DataFrames.TransformationResult}, idx_agg::Base.RefValue{Union{Nothing, Vector{Int64}}})
       @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:466
     [5] macro expansion
       @ ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:556 [inlined]
     [6] (::DataFrames.var"#1871#1877"{GroupedDataFrame{DataFrame}, Bool, Bool, DataFrame, Dict{Symbol, Tuple{Bool, Int64}}, Vector{DataFrames.TransformationResult}, Base.RefValue{Union{Nothing, Vector{Int64}}}, Bool, Pair{Int64, Pair{var"#21#22", Symbol}}})()
       @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:499
```

Reply:
OK - thank you!

Reply:
Actually with `@sync` the error is better:
```julia
julia> f(x) = nonexistent(x)
f (generic function with 1 method)

julia> combine(gd, :x => (x -> f(x)))
ERROR: TaskFailedException

    nested task error: UndefVarError: nonexistent not defined
    Stacktrace:
     [1] f(x::SubArray{Int64, 1, Vector{Int64}, Tuple{Vector{Int64}}, false})
       @ Main ./REPL[115]:1
     [2] #145
       @ ./REPL[116]:1 [inlined]
     [3] do_call(f::var"#145#146", idx::Vector{Int64}, starts::Vector{Int64}, ends::Vector{Int64}, gd::GroupedDataFrame{DataFrame}, incols::Tuple{Vector{Int64}}, i::Int64)
       @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/callprocessing.jl:95
     [4] _combine_process_pair(cs_i::Pair, optional_i::Bool, parentdf::DataFrame, gd::GroupedDataFrame{DataFrame}, seen_cols::Dict{Symbol, Tuple{Bool, Int64}}, trans_res::Vector{DataFrames.TransformationResult}, idx_agg::Base.RefValue{Union{Nothing, Vector{Int64}}})
       @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:466
     [5] macro expansion
       @ ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:556 [inlined]
     [6] (::DataFrames.var"#3079#3085"{GroupedDataFrame{DataFrame}, Bool, Bool, DataFrame, Dict{Symbol, Tuple{Bool, Int64}}, Vector{DataFrames.TransformationResult}, Base.RefValue{Union{Nothing, Vector{Int64}}}, Bool, Pair{Int64, Pair{var"#145#146", Symbol}}})()
       @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:499
Stacktrace:
 [1] sync_end(c::Channel{Any})
   @ Base ./task.jl:364
 [2] macro expansion
   @ ./task.jl:383 [inlined]
 [3] _combine(gd::GroupedDataFrame{DataFrame}, cs_norm::Vector{Any}, optional_transform::Vector{Bool}, copycols::Bool, keeprows::Bool, renamecols::Bool)
   @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:542
 [4] _combine_prepare(gd::GroupedDataFrame{DataFrame}, cs::Union{Regex, AbstractString, Function, Signed, Symbol, Unsigned, Pair, AbstractVector{T} where T, Type, All, Between, Cols, InvertedIndex}; keepkeys::Bool, ungroup::Bool, copycols::Bool, keeprows::Bool, renamecols::Bool)
   @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:66
 [5] #combine#529
   @ ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:645 [inlined]
 [6] combine(gd::GroupedDataFrame{DataFrame}, cs::Pair{Symbol, var"#145#146"})
   @ DataFrames ~/.julia/dev/DataFrames/src/groupeddataframe/splitapplycombine.jl:647
 [7] top-level scope
   @ REPL[116]:1
```

Reply:
OK

Reply:
Actually given that `@sync` changes the type of the thrown exception, I suggest we go with the approach at https://github.com/JuliaData/DataFrames.jl/pull/2574#discussion_r541750910 instead. It's quite verbose, but the exception type is correct so we won't have to break that in the future, and the full backtrace is printed in the last exception. I've pushed a commit to do that.